File: packages/ui-flutter/lib/src/message_bubble.dart
import 'package:flutter/widgets.dart';

import 'app_text.dart';
import 'theme.dart';

enum KinetixMessageVariant { sent, received }

enum KinetixMessageStatus { sent, delivered, read }

/// Mirrors `packages/ui/src/components/message-bubble.tsx`: a
/// sent/received chat bubble with grouping, timestamp, and a status
/// tick, plus a typing indicator. `grouped` reduces the outer top
/// corner's radius via [BorderRadius.only] — the same precise
/// single-corner treatment as the web/Compose ports (unlike the SwiftUI
/// port, which falls back to a uniform radius since this package
/// couldn't verify `UnevenRoundedRectangle`'s OS-version floor without a
/// local toolchain). No icon library wired in — the status ticks are
/// plain "✓"/"✓✓" glyphs.
class KinetixMessageBubble extends StatelessWidget {
  const KinetixMessageBubble(
    this.text, {
    super.key,
    this.variant = KinetixMessageVariant.received,
    this.timestamp,
    this.status,
    this.grouped = false,
    this.avatar,
  });

  final String text;
  final KinetixMessageVariant variant;
  final String? timestamp;
  final KinetixMessageStatus? status;
  final bool grouped;
  final Widget? avatar;

  @override
  Widget build(BuildContext context) {
    final c = KinetixTheme.of(context);
    final sent = variant == KinetixMessageVariant.sent;

    const radius = Radius.circular(16);
    const smallRadius = Radius.circular(8);
    final borderRadius = BorderRadius.only(
      topLeft: !sent && grouped ? smallRadius : radius,
      topRight: sent && grouped ? smallRadius : radius,
      bottomLeft: radius,
      bottomRight: radius,
    );

    final bubble = Container(
      constraints: const BoxConstraints(maxWidth: 280),
      padding: const EdgeInsets.symmetric(horizontal: 14, vertical: 8), // px-3.5/spacing-2, off-scale
      decoration: BoxDecoration(color: sent ? c.primary : c.muted, borderRadius: borderRadius),
      child: Text(text, style: AppText.bodySm.copyWith(color: sent ? c.primaryForeground : c.foreground)),
    );

    final Widget? meta = (timestamp != null || (sent && status != null))
        ? Padding(
            padding: const EdgeInsets.symmetric(horizontal: 4),
            child: Row(
              mainAxisSize: MainAxisSize.min,
              children: [
                if (timestamp != null) Text(timestamp!, style: AppText.labelSm.copyWith(color: c.mutedForeground)),
                if (sent && status != null) ...[
                  if (timestamp != null) const SizedBox(width: 4),
                  Text(
                    status == KinetixMessageStatus.sent ? '✓' : '✓✓',
                    style: AppText.labelSm.copyWith(
                      color: status == KinetixMessageStatus.read ? c.primary : c.mutedForeground,
                    ),
                  ),
                ],
              ],
            ),
          )
        : null;

    return Row(
      mainAxisSize: MainAxisSize.min,
      crossAxisAlignment: CrossAxisAlignment.end,
      children: [
        if (!sent && avatar != null) Padding(padding: const EdgeInsets.only(right: 8), child: avatar),
        Flexible(
          child: Column(
            crossAxisAlignment: sent ? CrossAxisAlignment.end : CrossAxisAlignment.start,
            mainAxisSize: MainAxisSize.min,
            children: [
              bubble,
              if (meta != null) ...[const SizedBox(height: 4), meta],
            ],
          ),
        ),
      ],
    );
  }
}

class KinetixTypingIndicator extends StatefulWidget {
  const KinetixTypingIndicator({super.key});

  @override
  State<KinetixTypingIndicator> createState() => _KinetixTypingIndicatorState();
}

class _KinetixTypingIndicatorState extends State<KinetixTypingIndicator> with SingleTickerProviderStateMixin {
  late final AnimationController _controller;

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(vsync: this, duration: const Duration(milliseconds: 1200))..repeat();
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final c = KinetixTheme.of(context);
    return Container(
      padding: const EdgeInsets.symmetric(horizontal: 14, vertical: 10),
      decoration: BoxDecoration(color: c.muted, borderRadius: BorderRadius.circular(16)),
      child: Row(
        mainAxisSize: MainAxisSize.min,
        children: [
          for (var i = 0; i < 3; i++) ...[
            if (i > 0) const SizedBox(width: 4),
            AnimatedBuilder(
              animation: _controller,
              builder: (context, _) {
                final phase = (i * 150) / 1200;
                final t = (_controller.value + phase) % 1.0;
                final bounce = t < 0.3 ? t / 0.3 : (1 - t) / 0.7;
                return Transform.translate(
                  offset: Offset(0, -3 * bounce),
                  child: Container(
                    width: 6,
                    height: 6,
                    decoration: BoxDecoration(shape: BoxShape.circle, color: c.mutedForeground),
                  ),
                );
              },
            ),
          ],
        ],
      ),
    );
  }
}

File: packages/ui-swiftui/Sources/KinetixUI/MessageBubble.swift
//
// MessageBubble.swift — KinetixMessageBubble / KinetixTypingIndicator.
//
// Mirrors packages/ui/src/components/message-bubble.tsx: a sent/received
// chat bubble with grouping, timestamp, and a status tick, plus a typing
// indicator. `grouped` uses a uniform (not single-corner) radius
// reduction as its consecutive-run cue — `UnevenRoundedRectangle`'s exact
// OS-version floor isn't certain enough to rely on without a local
// toolchain to verify against this package's iOS 16.0 floor, so this is
// a documented, simpler fallback rather than the web/Compose ports'
// precise single-corner treatment. SF Symbols give a real checkmark
// glyph, but no "double checkmark" one exists, so delivered/read use two
// plain "✓" characters, matching the web/Compose "✓✓" convention.
//

import SwiftUI

public enum KinetixMessageVariant {
    case sent, received
}

public enum KinetixMessageStatus {
    case sent, delivered, read
}

public struct KinetixMessageBubble<Avatar: View>: View {
    @Environment(\.kinetixColors) private var colors

    private let text: String
    private let variant: KinetixMessageVariant
    private let timestamp: String?
    private let status: KinetixMessageStatus?
    private let grouped: Bool
    private let avatar: Avatar

    public init(
        _ text: String,
        variant: KinetixMessageVariant = .received,
        timestamp: String? = nil,
        status: KinetixMessageStatus? = nil,
        grouped: Bool = false,
        @ViewBuilder avatar: () -> Avatar = { EmptyView() }
    ) {
        self.text = text
        self.variant = variant
        self.timestamp = timestamp
        self.status = status
        self.grouped = grouped
        self.avatar = avatar()
    }

    private var sent: Bool { variant == .sent }

    public var body: some View {
        HStack(alignment: .bottom, spacing: 8) { // spacing/2
            if !sent {
                avatar
            } else {
                Spacer(minLength: 40)
            }

            VStack(alignment: sent ? .trailing : .leading, spacing: 4) {
                Text(text)
                    .font(.kinetixBodySm)
                    .foregroundStyle(sent ? colors.actionForeground : colors.foreground)
                    .padding(.horizontal, 14) // px-3.5, off-scale — same call as KinetixBadge
                    .padding(.vertical, 8) // spacing/2
                    .background(
                        sent ? colors.action : colors.muted,
                        in: RoundedRectangle(cornerRadius: grouped ? 10 : 16, style: .continuous)
                    )
                    .frame(maxWidth: 280, alignment: sent ? .trailing : .leading)

                if timestamp != nil || (sent && status != nil) {
                    HStack(spacing: 4) {
                        if let timestamp {
                            Text(timestamp).font(.kinetixLabelSm).foregroundStyle(colors.mutedForeground)
                        }
                        if sent, let status {
                            Text(status == .sent ? "✓" : "✓✓")
                                .font(.kinetixLabelSm)
                                .foregroundStyle(status == .read ? colors.action : colors.mutedForeground)
                        }
                    }
                    .padding(.horizontal, 4)
                }
            }

            if !sent {
                Spacer(minLength: 40)
            }
        }
    }
}

public struct KinetixTypingIndicator: View {
    @Environment(\.kinetixColors) private var colors
    @State private var animate = false

    public init() {}

    public var body: some View {
        HStack(spacing: 4) { // spacing/1
            ForEach(0..<3, id: \.self) { i in
                Circle()
                    .fill(colors.mutedForeground)
                    .frame(width: 6, height: 6)
                    .offset(y: animate ? -3 : 0)
                    .animation(
                        .easeInOut(duration: 0.6).repeatForever(autoreverses: true).delay(Double(i) * 0.15),
                        value: animate
                    )
            }
        }
        .padding(.horizontal, 14) // px-3.5, off-scale
        .padding(.vertical, 10) // spacing/2.5, off-scale
        .background(colors.muted, in: RoundedRectangle(cornerRadius: 16, style: .continuous))
        .onAppear { animate = true }
    }
}
